### PR TITLE
Add sharing map documentation and move unit test sections

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 088418d9b6036ba701b7d5292c503a982a05e381).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112139627

File: unit/util/sharing_map.cpp [hunk 1]
```diff
@@ -176,94 +176,7 @@ void sharing_map_internals_test()
...
-
-        REQUIRE(delta_view[1].k == 0);
-        REQUIRE(delta_view[1].m == "c");
-        REQUIRE(delta_view[1].get_other_map_value() == "a");
-      }
-    }
-  }
-
-  SECTION("two maps shape, sharing")
+  SECTION("delta view (sharing, one of the maps is deeper")
```
Comment (tautschnig) [inline unit/util/sharing_map.cpp]: Missing closing `)`

File: src/util/sharing_map.h [hunk 2]
```diff
@@ -535,6 +535,22 @@ class sharing_mapt
     return lp;
   }
 
+  /// Move a container node (containing a single leaf) further down the tree
+  /// such as to resolve a collision with another key-value pair. This method is
+  /// called by insert() to resolve a collision between a key-value pair to be
+  /// newly inserted, and a key-value pair existing in the map.
+  ///
+  /// \param i the depth of the inner node pointing to the container node with
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Coding guidelines recommend a colon after the parameter name. Speaking of which: `i` really isn't a good name, the only acceptable place for such a short name is a loop index IMHO.

File: src/util/sharing_map.h [hunk 3]
```diff
@@ -535,6 +535,22 @@ class sharing_mapt
...
+  ///   a single leaf
+  /// \param key_suffix hash code of the existing key in the map, shifted to the
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Colon after parameter name.

File: src/util/sharing_map.h [hunk 4]
```diff
@@ -535,6 +535,22 @@ class sharing_mapt
...
+  ///   right by `chunk * i` bits (i.e., `key_suffix` is the rest of the hash
+  ///   code used to determine the position of the key-value pair below level
+  ///   `i`
+  /// \param bit_last last portion of the hash code of the key existing in the
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Colon

File: src/util/sharing_map.h [hunk 5]
```diff
@@ -535,6 +535,22 @@ class sharing_mapt
...
+  ///   map (inner[bit_last] points to the container node to move further down
+  ///   the tree)
+  /// \param inner inner node of which the child inner[bit_last] is the
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Colon

File: src/util/sharing_map.h [hunk 6]
```diff
@@ -535,6 +535,22 @@ class sharing_mapt
...
+  ///   container node to move further down the tree
```
Comment (tautschnig) [inline src/util/sharing_map.h:553]: Missing `\return ...`

File: src/util/sharing_map.h [hunk 7]
```diff
@@ -545,6 +561,20 @@ class sharing_mapt
     const innert &n,
     std::function<void(const key_type &k, const mapped_type &m)> f) const;
 
+  /// Add a delta item to the delta view if the value in the `container` (which
```
Comment (tautschnig) [inline src/util/sharing_map.h]: I prefer `\p container` to make clear you are referring to a parameter (there's no actual difference in rendering at the moment, though).

File: src/util/sharing_map.h [hunk 8]
```diff
@@ -545,6 +561,20 @@ class sharing_mapt
...
+  /// when a container containing a single leaf is encountered in the first map,
+  /// and the corresponding node in the second map is an inner node.
+  ///
+  /// \param container container node containing a single leaf, part of the
+  ///   first map in a call `map1.get_delta_view(map2, ...)`
+  /// \param inner inner node which is part of the second map
+  /// \param level depth of the nodes in the maps (both `container` and `inner`
+  ///   must be at the same depth in their respective maps)
+  /// \param delta_view delta view to add delta items to
+  /// \param only_common flag indicating if only items are added to the delta
```
Comment (tautschnig) [inline src/util/sharing_map.h]: For all of the above: missing colon after the parameter name.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4db3e6297c3571ffa858600b8430ff81b01ebc52).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112321942

**Pull request merged**